storage/badger/operation/common.go

```diff
@@ -521,6 +521,50 @@ func traverse(prefix []byte, iteration iterationFunc) func(*badger.Txn) error {
 	}
 }
 
+// findHighestAtOrBelow searches for the highest key with the given prefix and a height
+// at or below the target height, and retrieves and decodes the value associated with the
+// key into the given entity. The actual height of the key that is found
+// (which can be different from the target height) is stored in the actualHeight variable.
+// If no key is found, the function returns storage.ErrNotFound.
+func findHighestAtOrBelow(
+	prefix []byte,
+	height uint64,
+	entity interface{},
+	actualHeight *uint64,
```

Comment:
This seems very strange.  why not just return the actual height along with the error?

Also, why return a function / why not just have something like

```
func findHighestAtOrBelow(
  tx *badger.Txn,
  prefix []byte,
  height uint64,
  entity interface{},
) (
  uint64,
  error,
) {
    ...
}
```

Reply:
Looking a bit more into it, this code is like this because it is the encouraged pattern from badger db https://dgraph.io/docs/badger/get-started/#transactions.

Reply:
I see.  Looks like badger only support one-shot (e.g., blockchain-style) transactions rather than general/interactive transaction, hence the api is design in such a way that you must specify the full transaction within the `func(txn) {...}` body.